### DISCUSSION TITLE
How to share environment variables between WSL2 and host Windows 10?

### DISCUSSION THREAD

**Discussion by krave1986** [microsoft/WSL#7725] [Q&A]
I am looking for some way to share environment between WSL2 and host machine - Windows 10.
I checked [this article](https://devblogs.microsoft.com/commandline/share-environment-vars-between-wsl-and-windows/) until I found it is WSL related instead of WSL2.

Do you have any advices how to do this in WSL2 ?

Thank you!

**Comment by Biswa96** [comment]
AFAIK, it is same in WSL2.

**Comment by krave1986** [comment]
@Biswa96 Thank you for your prompt reply.

I have tried:

1. Run `setx WSLENV "$env:WSLENV:ONEDRIVE"/p` from PowerShell which is the same syntax listed in the article I mentioned earlier. But it turns out the syntax is not valid. 
![syntaxError](https://user-images.githubusercontent.com/16416541/142479843-10f9b42a-c86b-4034-9e44-4f14f0f26f32.png)

2. Run `export WSLENV=HOME/p` from WSL2 Ubuntu disto. The HOME variable is only available within `cmd.exe` which is invoked directly from the session in which I exported the variable. It is not available in other session.

I have no idea about what have I done wrong? My purpose is to access $OneDrive from within Ubuntu in order to bind mount it into Docker container.

**Comment by Biswa96** [comment]
What I understand is that you want to pass an environment variable from Windows to WSL. Then the correct flag might be `/u`. Here is an example to pass `%UserProfile%`.

1. In Command Prompt:

```cmd
C:\> set WSLENV=USERPROFILE/u
```

2. In WSL, the result will be like this :

```sh
$ env | grep -i user
USERPROFILE=C:\Users\x
USER=x
WSLENV=USERPROFILE/u
```

**Comment by krave1986** [comment]
@Biswa96 I can confirm that your solution works but constrained in one single session.

For example, open cmd.exe first, set WSLENV as you did. I have to run `ubuntu` right away to enter WSL2. That way $USERPROFILE can be accessed normally. By the way, I do need `/p` flag because I need path translation between Windows and WSL2.
![正常](https://user-images.githubusercontent.com/16416541/142495230-03788263-80f4-49de-80c9-d02405af043f.png)

But if I open another terminal , I can no loager access the variable. No need to say availability across system reboot.
![异常](https://user-images.githubusercontent.com/16416541/142495485-1bc3db89-0e9b-4af4-b7a9-66e57560a36a.png)

Maybe any suggestion about the variable persistence?

On the other hand, we can't set WSLENV from PowerShell, can we?

**Comment by Biswa96** [comment]
> But if I open another terminal , I can no loager access the variable.

Mine was just an example. To make the variable persistant, you need to use setx command or control panel's advanced system settings. The article already mentioned that.

**Comment by krave1986** [comment]
Here are what I have found today.

1. The setx syntax used in the article is not correct.
PowerShell interprets `/p` in `setx WSLENV "$env:WSLENV:ONEDRIVE"/p` as password option which is not we expect.
At least not in PowerShell on Windows. I tried both version 6 and version 7.

2. The value that WSL2 expects is environment variable already defined. WSL2 will try to use value of WSLENV to match what already been defined. If anything matches, the matched environment will be available inside Ubuntu. So for example, I have `OneDrive` variable already defined on Windows.

The correct way to share `OneDrive` with WSL2 is:

`setx WSLENV OneDrive/p`

No double qoutes around the variable allowed here or include entire `OneDrive/p` into double qoutes because of point 1 above.

3. The article metioned that the value WSLENV on Windows is a semicolon-delimited list which is not true. It should also be a a colon-delimited list.

So if I want to share both `OneDrive` and `DEVRESOURCES` with WSL2 , it should be
`setx WSLENV OneDrive/p:DEVRESOURCES/p`

But I still have trouble when I want to append new variable to existing `WSLENV`.
For example, if I want to append `TEMP` to `WSLENV`, I tried:

`setx WSLENV "$($env:WSLENV):TEMP/p"`

It will add prefiex `WT_SESSION:` and suffix `:WT_PROFILE_ID` into the value.


![多出了前后缀](https://user-images.githubusercontent.com/16416541/142694029-6a16f3b9-6609-4ed8-ad60-c097f9428f06.png)

Any suggestions?

**Comment by mortenrs** [comment]
To set it in Powershell use:

```ps1
[Environment]::SetEnvironmentVariable("WSLENV", $env:WSLENV + "USERPROFILE/p:", [System.EnvironmentVariableTarget]::User)
```
